linera-core/src/chain_worker/state.rs

```diff
@@ -546,24 +552,52 @@ where
     pub(super) async fn query_application(
         &mut self,
         query: Query,
-        incoming_execution_requests: futures::channel::mpsc::UnboundedReceiver<ExecutionRequest>,
-        runtime_request_sender: std::sync::mpsc::Sender<ServiceRuntimeRequest>,
     ) -> Result<Response, WorkerError> {
         self.0.ensure_is_active()?;
+        self.prepare_to_query_application();
         let local_time = self.0.storage.clock().current_time();
         let response = self
             .0
             .chain
             .query_application(
                 local_time,
                 query,
-                incoming_execution_requests,
-                runtime_request_sender,
+                &mut self.0.execution_state_receiver,
+                &mut self.0.runtime_request_sender,
             )
             .await?;
         Ok(response)
     }
 
+    /// Configures the [`QueryContext`] before executing a service to handle a query.
+    ///
+    /// Restarts the service runtime actor if needed, otherwise just updates the local time of the
+    /// context.
+    fn prepare_to_query_application(&mut self) {
+        let new_context = self.0.current_query_context();
+
+        let mut expected_context = new_context;
+        if let Some(old_context) = self.0.last_query_context {
+            expected_context.local_time = old_context.local_time;
+        }
+
+        let request = if self.0.last_query_context != Some(expected_context) {
+            ServiceRuntimeRequest::ChangeContext {
+                context: new_context,
+            }
+        } else {
+            ServiceRuntimeRequest::UpdateLocalTime {
+                local_time: new_context.local_time,
+            }
+        };
```

Comment:
Is there any issue with passing the context and the local time every time with the query?

Reply:
I don't think so. The context causes the runtime (and hence the service) to restart, falling back to the previous behavior. I thought the local time would affect the cache inside the `ServiceRuntime` in the SDK, but what is cached is the system time (i.e., the block time) instead. I think it would affect oracles, but it shouldn't matter in the service I think :thinking:

Reply:
Generally, we probably want to minimize the number of commands.

nit: If the runtime is restarted by `ChangeContext` maybe the name could hint at that.

Reply:
Done.